torchgen/gen.py

```diff
@@ -2041,22 +2041,55 @@ def key_func(
 
     # We need to easily map from [inplace_op_name] -> [functional_op] for the functionalization pass,
     # so here I generate a mapping from every operator name to its corresponding functional NativeFunction (if it exist).
+    def group_native_fns_by_mutability(
+        native_functions: Sequence[NativeFunction],
+    ) -> Dict[FunctionSchema, List[NativeFunction]]:
+        pre_grouped_native_functions: Dict[
+            FunctionSchema, List[NativeFunction]
+        ] = defaultdict(list)
+        for f in native_functions:
+            d = pre_grouped_native_functions[f.func.mutation_agnostic_signature()]
+            d.append(f)
+        return pre_grouped_native_functions
+
     pre_grouped_d: Dict[
-        FunctionSchema, Dict[SchemaKind, NativeFunction]
-    ] = pre_group_native_functions(native_functions)
-    to_functional_op: Dict[OperatorName, Optional[NativeFunction]] = {
-        k: v
-        for d in [
-            {
-                f.func.name: pre_grouped_d[func][SchemaKind.functional]
-                if SchemaKind.functional in pre_grouped_d[func].keys()
-                else None
-                for f in pre_grouped_d[func].values()
-            }
-            for func in pre_grouped_d.keys()
+        FunctionSchema, List[NativeFunction]
+    ] = group_native_fns_by_mutability(native_functions)
+    to_functional_op: Dict[OperatorName, Optional[NativeFunction]] = defaultdict(None)
+    for schema, funcs in pre_grouped_d.items():
+        # We can't use schemaKind() to detect mutability, because we have some mutable ops
+        # that are neither inplace nor out= ops (because they have mutable non-self, positional args)
```

Comment:
Can we add another enum to schemaKind then? Ideally functional ops would also go in GroupedNativeFunctions. This might help shorten this code.

Reply:
Hmm it seems to me like it would hurt more than it would help.

We could add a new `SchemaKind` to represent "mutable ops that are neither out= nor inplace", but then we'd need to update every other piece of codegen that cares `SchemaKind` to care about the new kind.

We can't get any of this into `NativeFunctionsGroup` very easily, since we requires a group to have an out= variant - but `_fused_moving_avg_obs_fq_helper` is an example of an op that should have a mutable + functional version, but doesn't actually have an out= or inplace variant.

I figured that since this "mutable -> functional" mapping isn't really important anywhere outside of functionalization, then the one-off mapping I'm constructing here is less work than trying to plumb new info through the rest of the codegen. But happy to change it if you have a different opinion.

Reply:
> We can't get any of this into NativeFunctionsGroup very easily, since we requires a group to have an out= variant - but _fused_moving_avg_obs_fq_helper is an example of an op that should have a mutable + functional version, but doesn't actually have an out= or inplace variant.

This poses the question: should it have an out variant? It seems easy enough to take the mutable version and tack on an out argument to make it have an out variant to. Mobile would certainly appreciate it. Would that help?

> We could add a new SchemaKind to represent "mutable ops that are neither out= nor inplace", but then we'd need to update every other piece of codegen that cares SchemaKind to care about the new kind.

I'm not sure this is necessarily true, but I'd have to dig around to see what the implications are.

Reply:
>  should it have an out variant? ... Would that help?

Oh that's fair, that would help. At least, it would let me use schema kinds to do the grouping in this code (I'll make that change).

We still need to live with a side-car / custom mapping from mutable -> functional ops though for the functionalization codegen though, unless we're willing to force all mutable ops to have an out= variant in order to go through functionalization (which I don't think we are? - `at::zero_out` would be... very useless haha)

Reply:
> We still need to live with a side-car / custom mapping from mutable -> functional ops though for the functionalization codegen though, unless we're willing to force all mutable ops to have an out= variant in order to go through functionalization (which I don't think we are? - at::zero_out would be... very useless haha)

To recap for a moment why it would be useless. We started this affair with `zero_`, and completing out the signatures gives us this:

```
zero_(Tensor(a!) self) -> Tensor(a!) self
zero(Tensor self) -> Tensor
zero_out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```

`zero` is useful because it is basically `zeros_like` (it's not as good because `zeros_like` allows more customization for the output dtype, etc; see also what @cpuhrsch has been advocating for.) `zero_out` is useless because in this operator, the input tensor is completely ignored, so the inplace variant suffices. (This is not *entirely* true, since the out tensor will get resized to match self).

I feel like it would be easier to make the call that we should make `zero_out` if we didn't actually have to write dozens of lines of boilerplate to define it. How many more of these do you think we'd have to add?

Reply:
From a quick test from poking the codegen, I found:

Ops that only have an `inplace` variant: 37 (technically 83 if you include foreach ops)

Ops that have an `inplace` and `functional`, but not an `out=` variant: 52

Ops that have only an `out=` variant (no functional though): 27

Reply:
my vote is tentatively for forcing all operators to have out= variants but I'm not the one doing the work. You may want to shop around for other opinions. Now that torchgen is publicly visible it's also a good time to consider rewriting the input yaml format.